### PR TITLE
feat: add koa-pagination

### PR DESCRIPTION
**Pull request by wangsijie**
## Summary

Add middleware `koaPagination` to handle pagination in both request and response.

1. Parse page-related query: `page`, `per_page`，can set default value. Wrapped into `context.pagination`.
2. After data is fetched and handled properly, set total count to `context.pagination.count`, this middleware will then set response headers: customized `Total-Page` and standard `Link` (https://datatracker.ietf.org/doc/html/rfc5988).

## Testing

`src/middleware/koa-pagination.test.ts`
`src/utils/pagination.test.ts`

### CODE REVIEW

File: packages/core/src/middleware/koa-pagination.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import { MiddlewareType } from 'koa';
+import { number } from 'zod';
+
+import RequestError from '@/errors/RequestError';
+import { buildLink } from '@/utils/pagination';
+
+export interface Pagination {
+  page: number; // Begin from 1
+  size: number;
```
Comment (xiaoyijun) [inline packages/core/src/middleware/koa-pagination.ts:9]: use `perPage` to keep aligning with link?

File: packages/core/src/middleware/koa-pagination.test.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    });
+    expect(setHeader).toHaveBeenCalledWith('Total-Number', '100');
+  });
+
+  describe('Link in response header', () => {
+    it('should append 2 times in 1 of 1 (first, last)', async () => {
+      const context = createContext({});
+      await koaPagination({ defaultPageSize: 20 })(context, async () => {
+        // eslint-disable-next-line @silverhand/fp/no-mutation
+        context.pagination.count = 10;
```
Comment (simeng-li) [inline packages/core/src/middleware/koa-pagination.test.ts:91]: nit: if it is applicable could mock sth like

 ```js
    ctx  = {
       Link: net Set(),
       append: jest.fn((key:string, value: string) => {
         ctx[key].add(value)
       }
```

might need some type restrictions and mock clears additional work as well. Anyway could test the first, pre, next, last, more precisely.

**Review by simeng-li — COMMENTED** [review]
please check

File: packages/core/src/middleware/koa-pagination.ts (hunk 1 above)
Comment (wangsijie) [reply packages/core/src/middleware/koa-pagination.ts]: use `limit` instead

File: packages/core/src/middleware/koa-pagination.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+}: PaginationConfig = {}): MiddlewareType<StateT, WithPaginationContext<ContextT>, ResponseBodyT> {
+  return async (ctx, next) => {
+    try {
+      // Query values are all string, need to convert to number first.
+      ctx.pagination = {
+        page: ctx.request.query.page
+          ? number().positive().parse(Number(ctx.request.query.page))
+          : 1,
+        size: ctx.request.query.per_page
+          ? number().positive().max(maxPageSize).parse(Number(ctx.request.query.per_page))
```
[the remark this answers is not in the document]
Comment (wangsijie) [reply packages/core/src/middleware/koa-pagination.ts]: `size` will break lint and seems not recommended: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/v34.0.1/docs/rules/explicit-length-check.md

Use `page_size` instead.

File: packages/core/src/middleware/koa-pagination.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    // Only handle response when count value is set.
+    if (ctx.pagination.totalCount !== undefined) {
+      const { limit, offset, totalCount } = ctx.pagination;
+      const totalPage = Math.ceil(totalCount / limit) || 1; // Minimum page number is 1
+
+      // Our custom response header: Total-Number
+      ctx.set('Total-Number', String(totalCount));
+
+      // Response header's `Link`: https://datatracker.ietf.org/doc/html/rfc5988
+      const page = Math.floor(offset / limit) + 1; // Start from 1
```
Comment (simeng-li) [inline packages/core/src/middleware/koa-pagination.ts:55]: nit: looks like `page` is declared twice (one on query params). could use pageIndex here.

**Review by simeng-li — APPROVED** [review]
LGTM

File: packages/core/src/middleware/koa-pagination.ts (hunk 4 above)
Comment (wangsijie) [reply packages/core/src/middleware/koa-pagination.ts:55]: the one on query params has been renamed before : )

**Pull request merged**